apps/api/src/services/portfolio-snapshotter.service.ts

```diff
@@ -62,26 +64,151 @@ export class PortfolioSnapshotter {
+      repositoryLogger.debug(
+        `[PortfolioSnapshotter] No balances found for agent ${agentId}, skipping snapshot`,
+      );
+      return;
+    }
+
+    // Retry loop for the entire snapshot if there are any failures
+    let attemptNumber = 0;
+    const priceMap: Map<string, PriceReport | null> = new Map();
+
+    // Loop will run maxRetries + 1 times (initial attempt + retries)
+    while (attemptNumber < maxRetries + 1) {
+      attemptNumber++;
+
+      // Add exponential backoff delay between retry attempts (except first attempt)
+      if (attemptNumber > 1) {
+        // Exponential backoff: attemptNumber - 2 is intentional to get 1s, 2s, 4s progression
+        // Attempt 2: 2^(2-2) = 2^0 = 1 second
+        // Attempt 3: 2^(3-2) = 2^1 = 2 seconds
+        // Attempt 4: 2^(4-2) = 2^2 = 4 seconds
+        const backoffDelay = Math.min(
+          1000 * Math.pow(2, attemptNumber - 2),
+          10000,
+        ); // Max 10 seconds
+        repositoryLogger.debug(
+          `[PortfolioSnapshotter] Retry attempt ${attemptNumber}/${maxRetries + 1} for agent ${agentId} after ${backoffDelay}ms delay`,
+        );
+        await new Promise((resolve) => setTimeout(resolve, backoffDelay));
+      }
+
+      // Step 1: Determine which tokens need prices
+      const tokensNeedingPrices: string[] = [];
+      if (attemptNumber === 1) {
+        // First attempt: fetch all tokens
+        tokensNeedingPrices.push(...balances.map((b) => b.tokenAddress));
+      } else {
+        // Subsequent attempts: only fetch tokens that don't have prices yet
+        for (const balance of balances) {
+          if (
+            balance.amount > 0 &&
+            priceMap.get(balance.tokenAddress) === null
+          ) {
+            tokensNeedingPrices.push(balance.tokenAddress);
+          }
+        }
+      }
+
+      // If we have all prices already, we're done
+      if (tokensNeedingPrices.length === 0) {
+        repositoryLogger.debug(
+          `[PortfolioSnapshotter] All prices already fetched for agent ${agentId}`,
+        );
+        break;
+      }
+
+      // Step 2: Try batch pricing for tokens that need prices
+      repositoryLogger.debug(
+        `[PortfolioSnapshotter] Fetching prices for ${tokensNeedingPrices.length} tokens (attempt ${attemptNumber}/${maxRetries})`,
+      );
+      const newPrices =
+        await this.priceTracker.getBulkPrices(tokensNeedingPrices);
+
+      // Merge new prices into our map (preserving existing successful prices)
+      for (const [tokenAddress, priceReport] of newPrices) {
+        // Only update if we got a successful price (not null)
+        if (priceReport !== null) {
+          priceMap.set(tokenAddress, priceReport);
+        } else if (!priceMap.has(tokenAddress)) {
+          // Set to null if we don't have it yet (to track that we tried)
+          priceMap.set(tokenAddress, null);
+        }
+      }
+
+      // Step 3: Identify tokens that still need prices after batch
```

Comment:
I think we can simplify the logic here a bit.  Before the loop condition we can add a boolean `allPricesFetched` and default it to false, then add a check on `allPricesFetched` into the loop termination condition.  Then here we can just do a `find()` call on balances to check if there are *any* tokens left without a price, and if not, we can set `allPricesFetched` to true.  That simplifies the logic here by letting us get rid of the `failedTokens` array entirely.  It also means that when we exit the loop, we can check if `allPricesFetched` is true and if not, just immediately throw, rather than needing to re-check later if there are any prices missing.